src/plugins/intel_gna/transformations/insert_copy_layer.cpp

```diff
@@ -0,0 +1,328 @@
+            size_t bytesPerCropElement = 2;
+            size_t cropOffset = crop->offset.back() * bytesPerCropElement;
+            return (ALIGN64(cropOffset) != cropOffset);
+        }
+        return false;
+    }
+
+    // this not only mathematically trivial, has some WA for kaldi case
+    bool IsTrivialPermute(std::shared_ptr<ngraph::Node> node) {
+        auto permute = std::dynamic_pointer_cast<ngraph::opset8::Transpose>(node);
+        if (!permute) return false;
+
+        auto transpose_const = std::dynamic_pointer_cast<ngraph::op::Constant>(permute->input_value(1).get_node_shared_ptr());
+        if (!transpose_const) return false;
+
+        auto node_order = transpose_const->cast_vector<int64_t>();
+
+        if (node_order == std::vector<int64_t>({ 0, 3, 2, 1 })) {
+            return true;  // supported case
+        }
+
+        if (permute->get_input_size() == 0)
+            return false; // unsupported case
+
+        auto input = permute->input(0).get_source_output().get_node_shared_ptr();
+        auto input_order = permute->get_input_shape(0);
+        // auto inputs = layer->insData.begin()->lock();
+        // auto inputsOrder = inputs->getTensorDesc().getDims();
+        // cases when all permutations happened either between 1 and X shape where no other dims in between
+        auto permute_seq = genPermutations(node_order.begin(), node_order.end());
+        auto input_order_transformed = input_order;
+        for (auto && permute : permute_seq) {
+            // check dims of permuted
+            if (input_order_transformed[permute.first] == 1 &&
+                input_order_transformed[permute.second] == 1) {
+                return true;
+            }
+            if (input_order_transformed[permute.first] != 1 &&
+                input_order_transformed[permute.second] != 1) {
+                return false;
+            }
+            // check dims in between
+            for (int j = std::min(permute.first, permute.second) + 1; j < std::max(permute.first, permute.second); j++) {
+                if (input_order_transformed[j] != 1) {
+                    return false;
+                }
+            }
+            // apply permutation
+            std::swap(input_order_transformed[permute.first], input_order_transformed[permute.second]);
+        }
+        return true;
+    }
+
+    bool IsNonFunctionalGNANode(std::shared_ptr<ngraph::Node> node) {
+        return std::dynamic_pointer_cast<ngraph::opset8::Reshape>(node) ||
+               std::dynamic_pointer_cast<ngraph::opset8::Squeeze>(node) ||
+               std::dynamic_pointer_cast<ngraph::opset8::Unsqueeze>(node) ||
+               IsTrivialPermute(node);
+    }
+}// namespace
+
+HandleMultiConnectedLayerToConcat::HandleMultiConnectedLayerToConcat() {
+    MATCHER_SCOPE(HandleMultiConnectedLayerToConcat);
+
+    auto concat_op = ngraph::pattern::wrap_type<ngraph::opset8::Concat>();
```

Comment:
This way was suggested by Gleb
I also had a version that used the `run_on_function` and can switch back on it.
In `HandleMultiConnectedLayerToConcatAndMemory` we detect all `concat` and `memory` layers that go out of the layer, and this could be hard in backward traversing, but let me think about it.

Reply:
@AntonGrishin Ok, I see. If we keep these transformations using matchers, can we at least analyze possible merge of some of them? Looks like `HandleMultiConnectedLayerToConcat` and `InsertCopyBeforeConcatLayer` doing similiar things in terms of Concat detection, but have different triggers for insertion. From one point of view matchers should do small piece of work, but I worry about how they correlate with each other.